### PR TITLE
Revisit the unwrapping code in ControllerActionInvoker

### LINKED ISSUE #6679: Revisit the unwrapping code in ControllerActionInvoker
There's about a 150 lines of code we have in `ControllerActionInvoker` to massage the returned value to an `IActionResult`: https://github.com/aspnet/Mvc/blob/ccf9b2ec976147eb8452d8f3d50b446815765f48/src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs#L326-L412. Adding support for a new return type results in several dozens of lines of code and conditions making it pretty hard to manage and extend. We should see if this could be simplified.

Related item: The current code path performs reflection (https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs#L407-L411) to determine if a type is an `IActionResult`. We should change the code to avoid this.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #6679

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                {
+                    return Executors[i];
+                }
+            }
+
+            Debug.Fail("Should not get here");
+            throw new Exception();
+        }
+
+        // void LogMessgae(..)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs]: typeo

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            {
+                await (Task)executor.Execute(controller, arguments);
+                return new EmptyResult();
+            }
+
+            protected override bool CanConvert(ObjectMethodExecutor executor) => executor.MethodReturnType == typeof(Task);
+        }
+
+        // CustomAsync PerformActionAsync(..)
+        private class AsyncVoidResultExecutor : ActionMethodExecutor
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs]: Can we call this 'awaitable'? I don't want to confuse this with `async void`. Also if we have this case do we need `TaskResultExecutor`?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+            protected override bool CanConvert(ObjectMethodExecutor executor)
+            {
+                // Async method returning awaitable-of - IActionResult(e.g., Task<ViewResult>)
+                return executor.IsMethodAsync && typeof(IActionResult).IsAssignableFrom(executor.AsyncResultType);
+            }
+        }
+
+        // Task<object> GetPerson(..)
+        private class AsyncObjectResultExecutor : ActionMethodExecutor
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs]: Can we call this `Awaitable`?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 4]
```diff
@@ -321,74 +322,14 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
                     controller);
                 logger.ActionMethodExecuting(controllerContext, orderedArguments);
 
-                var returnType = executor.MethodReturnType;
-                if (returnType == typeof(void))
+                var converterResult = actionMethodExecutor.Execute(objectMethodExecutor, controller, orderedArguments);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: this needs a better name

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs (hunk 2 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs]: We can call `Execute` in `TaskResultExecutor` and avoid the cost of the `ExecuteAsync` pipeline.

**Comment by pranavkm** [comment]
🆙 📅 added some tests

**Pull request merged**